File: Data Structures & Algorithms/search-for-word-ii/submission-1.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.word = None


class Solution:
    def findWords(self, board, words):
        root = TrieNode()
        for word in words:
            node = root
            for char in word:
                if char not in node.children:
                    node.children[char] = TrieNode()
                node = node.children[char]
            node.word = word

        rows = len(board)
        cols = len(board[0])
        result = []

        def dfs(row, col, node):
            char = board[row][col]
            child = node.children.get(char)
            if child is None:
                return
            if child.word is not None:
                result.append(child.word)
                child.word = None
            board[row][col] = "#"
            if row > 0 and board[row - 1][col] != "#":
                dfs(row - 1, col, child)
            if row + 1 < rows and board[row + 1][col] != "#":
                dfs(row + 1, col, child)
            if col > 0 and board[row][col - 1] != "#":
                dfs(row, col - 1, child)
            if col + 1 < cols and board[row][col + 1] != "#":
                dfs(row, col + 1, child)
            board[row][col] = char
            if child.word is None and not child.children:
                node.children.pop(char)

        for row in range(rows):
            for col in range(cols):
                dfs(row, col, root)
        return result

File: Data Structures & Algorithms/search-for-word/submission-1.rs
impl Solution {
    pub fn exist(board: Vec<Vec<char>>, word: String) -> bool {
        let rows = board.len();
        let cols = board[0].len();
        let word: Vec<char> = word.chars().collect();
        if word.len() > rows * cols {
            return false;
        }

        let mut board = board;

        fn dfs(board: &mut [Vec<char>], word: &[char], row: i32, col: i32, index: usize) -> bool {
            if index == word.len() {
                return true;
            }
            if row < 0
                || row >= board.len() as i32
                || col < 0
                || col >= board[0].len() as i32
            {
                return false;
            }

            let row = row as usize;
            let col = col as usize;
            if board[row][col] != word[index] {
                return false;
            }

            let original = board[row][col];
            board[row][col] = '#';
            let found = dfs(board, word, row as i32 - 1, col as i32, index + 1)
                || dfs(board, word, row as i32 + 1, col as i32, index + 1)
                || dfs(board, word, row as i32, col as i32 - 1, index + 1)
                || dfs(board, word, row as i32, col as i32 + 1, index + 1);
            board[row][col] = original;
            found
        }

        for row in 0..rows {
            for col in 0..cols {
                if dfs(&mut board, &word, row as i32, col as i32, 0) {
                    return true;
                }
            }
        }
        false
    }
}